src/kademlia/dht_state.cpp

```diff
@@ -40,12 +40,40 @@ namespace dht
 {
 namespace
 {
-	node_id extract_node_id(bdecode_node const& e, string_view key)
+	node_ids_t extract_node_ids(bdecode_node const& e, string_view key)
 	{
-		if (e.type() != bdecode_node::dict_t) return node_id();
-		auto nid = e.dict_find_string_value(key);
-		if (nid.size() != 20) return node_id();
-		return node_id(nid);
+		if (e.type() != bdecode_node::dict_t) return node_ids_t();
+		node_ids_t ret;
+		// first look for an old-style nid
+		auto old_nid = e.dict_find_string_value(key);
+		if (old_nid.size() == 20)
+		{
+			ret.emplace_back(address(), node_id(old_nid));
+			return ret;
+		}
+		auto nids = e.dict_find_list(key);
+		if (!nids) return ret;
+		for (int i = 0; i < nids.list_size(); i++)
+		{
+			bdecode_node nid = nids.list_at(i);
+			if (nid.type() != bdecode_node::string_t) continue;
+			if (nid.string_length() < 20) continue;
+			char const* in = nid.string_ptr();
```

Comment:
on a separate note, looking at this, I wonder if there should be a way to get a ``span<char const>`` back here. I believe there's already a way to get a ``string_view``, but since this is not actually a string, and we parse stuff out from it, a ``span`` probably would make more sense.